File: src/lib/ipc/terminal.ts
import { Channel, invoke } from '@tauri-apps/api/core';

import type { TerminalAttachment, TerminalEvent, TerminalSession } from '$lib/domain/terminal';

export interface TerminalChannels {
	output: Channel<number[] | Uint8Array>;
	events: Channel<TerminalEvent>;
}

export interface TerminalApi {
	list(workspaceId: string): Promise<TerminalSession[]>;
	create(workspaceId: string, cols: number, rows: number): Promise<TerminalSession>;
	attach(
		workspaceId: string,
		terminalId: string,
		channels: TerminalChannels
	): Promise<TerminalAttachment>;
	detach(workspaceId: string, terminalId: string, subscriptionId: string): Promise<void>;
	write(workspaceId: string, terminalId: string, data: Uint8Array): Promise<void>;
	resize(
		workspaceId: string,
		terminalId: string,
		cols: number,
		rows: number
	): Promise<TerminalSession>;
	stop(workspaceId: string, terminalId: string): Promise<TerminalSession>;
	delete(workspaceId: string, terminalId: string): Promise<void>;
}

export const terminalApi: TerminalApi = {
	list: (workspaceId) => invoke('list_terminals', { workspaceId }),
	create: (workspaceId, cols, rows) => invoke('create_terminal', { workspaceId, cols, rows }),
	attach: (workspaceId, terminalId, channels) =>
		invoke('attach_terminal', {
			workspaceId,
			terminalId,
			outputChannel: channels.output,
			eventChannel: channels.events
		}),
	detach: (workspaceId, terminalId, subscriptionId) =>
		invoke('detach_terminal', { workspaceId, terminalId, subscriptionId }),
	write: (workspaceId, terminalId, data) =>
		invoke('write_terminal_input', { workspaceId, terminalId, data: Array.from(data) }),
	resize: (workspaceId, terminalId, cols, rows) =>
		invoke('resize_terminal', { workspaceId, terminalId, cols, rows }),
	stop: (workspaceId, terminalId) => invoke('stop_terminal', { workspaceId, terminalId }),
	delete: (workspaceId, terminalId) => invoke('delete_terminal', { workspaceId, terminalId })
};

export const createTerminalChannels = (
	onOutput: (data: Uint8Array) => void,
	onEvent: (event: TerminalEvent) => void
): TerminalChannels => {
	const output = new Channel<number[] | Uint8Array>();
	output.onmessage = (data) => onOutput(data instanceof Uint8Array ? data : Uint8Array.from(data));
	const events = new Channel<TerminalEvent>();
	events.onmessage = onEvent;
	return { output, events };
};

File: src-tauri/src/commands/terminal.rs
use std::sync::Arc;

use tauri::{ipc::Channel, State};

use crate::{
    domain::{
        terminal::{
            TerminalAttachment, TerminalEvent, TerminalId, TerminalSession, TerminalSubscriptionId,
        },
        workspace::WorkspaceId,
    },
    services::terminal::{
        TerminalCommandError, TerminalError, TerminalManager, TerminalSubscriber,
    },
};

struct ChannelSubscriber {
    output: Channel<Vec<u8>>,
    events: Channel<TerminalEvent>,
}

impl TerminalSubscriber for ChannelSubscriber {
    fn send_output(&self, data: Vec<u8>) -> Result<(), ()> {
        self.output.send(data).map_err(|_| ())
    }

    fn send_event(&self, event: TerminalEvent) -> Result<(), ()> {
        self.events.send(event).map_err(|_| ())
    }
}

#[tauri::command]
pub async fn list_terminals(
    workspace_id: WorkspaceId,
    manager: State<'_, TerminalManager>,
) -> Result<Vec<TerminalSession>, TerminalCommandError> {
    run_blocking(manager.inner().clone(), move |manager| {
        manager.list(&workspace_id)
    })
    .await
}

#[tauri::command]
pub async fn create_terminal(
    workspace_id: WorkspaceId,
    cols: u16,
    rows: u16,
    manager: State<'_, TerminalManager>,
) -> Result<TerminalSession, TerminalCommandError> {
    run_blocking(manager.inner().clone(), move |manager| {
        manager.create(workspace_id, cols, rows)
    })
    .await
}

#[tauri::command]
pub async fn attach_terminal(
    workspace_id: WorkspaceId,
    terminal_id: TerminalId,
    output_channel: Channel<Vec<u8>>,
    event_channel: Channel<TerminalEvent>,
    manager: State<'_, TerminalManager>,
) -> Result<TerminalAttachment, TerminalCommandError> {
    let subscriber = Arc::new(ChannelSubscriber {
        output: output_channel,
        events: event_channel,
    });
    run_blocking(manager.inner().clone(), move |manager| {
        manager.attach(&workspace_id, &terminal_id, subscriber)
    })
    .await
}

#[tauri::command]
pub async fn detach_terminal(
    workspace_id: WorkspaceId,
    terminal_id: TerminalId,
    subscription_id: TerminalSubscriptionId,
    manager: State<'_, TerminalManager>,
) -> Result<(), TerminalCommandError> {
    run_blocking(manager.inner().clone(), move |manager| {
        manager.detach(&workspace_id, &terminal_id, &subscription_id)
    })
    .await
}

#[tauri::command]
pub async fn write_terminal_input(
    workspace_id: WorkspaceId,
    terminal_id: TerminalId,
    data: Vec<u8>,
    manager: State<'_, TerminalManager>,
) -> Result<(), TerminalCommandError> {
    run_blocking(manager.inner().clone(), move |manager| {
        manager.input(&workspace_id, &terminal_id, &data)
    })
    .await
}

#[tauri::command]
pub async fn resize_terminal(
    workspace_id: WorkspaceId,
    terminal_id: TerminalId,
    cols: u16,
    rows: u16,
    manager: State<'_, TerminalManager>,
) -> Result<TerminalSession, TerminalCommandError> {
    run_blocking(manager.inner().clone(), move |manager| {
        manager.resize(&workspace_id, &terminal_id, cols, rows)
    })
    .await
}

#[tauri::command]
pub async fn stop_terminal(
    workspace_id: WorkspaceId,
    terminal_id: TerminalId,
    manager: State<'_, TerminalManager>,
) -> Result<TerminalSession, TerminalCommandError> {
    run_blocking(manager.inner().clone(), move |manager| {
        manager.stop(&workspace_id, &terminal_id)
    })
    .await
}

#[tauri::command]
pub async fn delete_terminal(
    workspace_id: WorkspaceId,
    terminal_id: TerminalId,
    manager: State<'_, TerminalManager>,
) -> Result<(), TerminalCommandError> {
    run_blocking(manager.inner().clone(), move |manager| {
        manager.delete(&workspace_id, &terminal_id)
    })
    .await
}

async fn run_blocking<T, F>(
    manager: TerminalManager,
    operation: F,
) -> Result<T, TerminalCommandError>
where
    T: Send + 'static,
    F: FnOnce(&TerminalManager) -> Result<T, TerminalError> + Send + 'static,
{
    tauri::async_runtime::spawn_blocking(move || operation(&manager))
        .await
        .map_err(|_| TerminalCommandError {
            code: "terminal_task_failed",
            message: "终端任务异常结束".into(),
        })?
        .map_err(Into::into)
}
